XMonad/Hooks/Modal.hs

```diff
@@ -0,0 +1,275 @@
+-- Portability :  unportable
+--
+-- This module implements modal keybindings for xmonad.
+--
+--------------------------------------------------------------------------------
+
+-- --< Imports & Exports >-- {{{
+module XMonad.Hooks.Modal
+  (
+ -- * Usage
+ -- $Usage
+    modal
+  , mode'
+  , mode
+  , Mode
+  , setMode
+  , exitMode
+ -- ** Provided Modes
+ -- $ProvidedModes
+  , noModModeLabel
+  , noModMode
+  , floatModeLabel
+  , floatMode
+  , overlayedFloatModeLabel
+  , overlayedFloatMode
+  , floatMap
+  , overlay
+ -- ** Logger
+  , logMode
+  ) where
+
+-- core
+import           XMonad
+
+-- base
+import           Data.Bits                      ( (.&.)
+                                                , complement
+                                                )
+import           Data.List
+import qualified Data.Map.Strict               as M
+import           XMonad.Actions.FloatKeys       ( keysMoveWindow
+                                                , keysResizeWindow
+                                                )
+import           XMonad.Prelude
+import qualified XMonad.Util.ExtensibleConf    as XC
+-- contrib
+import qualified XMonad.Util.ExtensibleState   as XS
+import           XMonad.Util.Grab
+import           XMonad.Util.Loggers
+
+-- }}}
+
+-- Original Draft By L.S.Leary : https://gist.github.com/LSLeary/6741b0572d62db3f0cea8e6618141b2f
+
+-- --< Usage >-- {{{
+
+-- $Usage
+--
+-- This module provides modal keybindings in xmonad. You can think of modes as
+-- submaps from 'XMonad.Actions.Submap', but after each action you execute, you
+-- land back into the submap until you explicitly exit the submap.
+-- To use this module you should apply the 'modal' function to the config, which
+-- will setup the list of modes (or rather, @XConfig Layout -> Mode@) your provide:
+--
+-- >
+-- > import XMonad
+-- > import XMonad.Hooks.Modal
+-- >
+-- > main :: IO ()
+-- > main =
+-- >   xmonad
+-- >     . modal [noModMode, floatMode 10, overlayedFloatMode 10, sayHelloMode]
+-- >     $ def
+-- >     `addiotnalKeysP` [ ("M-S-n", setMode noModModeLabel)
+-- >                      , ("M-S-r", setMode floatModeLabel)
+-- >                      , ("M-S-z", setMode overlayedFloatModeLabel)
+-- >                      ]
+-- >
+-- > sayHelloMode :: Mode
+-- > sayHelloMode = mode "Hello"
+-- >   $ const (M.fromList [((noModMask, xK_h), xmessage "Hello World! ")])
+--
+-- A 'Mode' has a label describing its purpose and keybinings (in form
+-- of @XConfig Layout -> M.Map (ButtonMask, KeySym) (X ())@). The label
+-- of the active mode can be logged with 'logMode' to be displayed in a
+-- status bar, for example (For more information check
+-- 'XMonad.Util.Loggers'). Check $ProvidedModes for a small collection of
+-- provided modes.
+
+-- }}}
+
+-- --< Types >-- {{{
+
+-- | The mode type. Use 'mode' or 'mode'' to create modes.
+data Mode = Mode
+  { label     :: String
+  , boundKeys :: XConfig Layout -> M.Map (ButtonMask, KeySym) (X ())
```

Comment:
I think the underlying type should be the same as the one in XConfig. Using EZConfig is a good idea, but I wanted to submit the first version without it so we get the other ergonomics right.

Reply:
Since we let users specify modes, I think having something like

``` haskell
  data BoundKeys
    = KeysEZ  (Map String (X ()))
    | KeysFun (XConfig Layout -> M.Map (ButtonMask, KeySym) (X ()))

  data Mode = Mode
    { label     :: String
    , boundKeys :: BoundKeys
    }
```

would perhaps be better.  We can really push the parsing of the keys out to `modal` (or `overlay`), since we don't really do anything with them internally.

I will try this out a bit this weekend so I can say more about the other ergonomics

Reply:
Sounds good :+1:

Reply:
Though, coming back to your point, it would be nice to somehow encode at the type level that we are using the internal representation in the actual state.  We could define `Mode` like

``` haskell
data Mode k = Mode
  { label     :: !String
  , boundKeys :: !k
  }
```

and then have two types along the lines of

``` haskell
data BoundKeys
  = KeysEZ  ![(String, X ())]
  | KeysFun !KeysIntern

newtype KeysIntern
  = KeysIntern (XConfig Layout -> M.Map (ButtonMask, KeySym) (X ()))
```

This along with a type class like

``` haskell
class BoundKeysClass k where
  getBoundKeys :: k -> XConfig Layout -> M.Map (ButtonMask, KeySym) (X ())

-- type BoundKeysClass :: Type -> Constraint
instance BoundKeysClass BoundKeys where
  getBoundKeys bk cnf = case bk of
    KeysEZ  k  -> mkKeymap cnf k
    KeysFun ki -> getBoundKeys ki cnf

instance BoundKeysClass KeysIntern where
  getBoundKeys = coerce
```

would probably allow us to put a `Mode KeysIntern` into the extensible state/config `newtype`s instead of just a `Mode`.  With a few helper functions it also wouldn't be too terrible ergonomic-wise for users.  Maybe this is over-engineering it though; just something I've thought of just now :)

Reply:
This looks super cool! Maybe something like that might be even more generally useful that it should warrant its own contrib module

Reply:
> This looks super cool! Maybe something like that might be even more generally useful that it should warrant its own contrib module

I can't think of another use-case right now, tbh, so it's probably fine to only use this inside this module for now (if we even want that!)

> I will try this out a bit this weekend so I can say more about the other ergonomics

Everything else seems totally fine to me!  I think in the docs we should mention that the default key to exit a mode is escape.  Also, there is no keybinding specified for the `sayHelloMode`, which might be confusing.

Reply:
> Using EZConfig is a good idea, but I wanted to submit the first version without it so we get the other ergonomics right.

There're still people who don't use EZConfig but might still want to use modal keybindings (sooner or later), like me. Last time I thought about EZConfig, I was discouraged by having to manually migrate lots of keybindings without the compiler telling me I made a mistake. And looking at its source, it seems like I wouldn't be told in runtime either. :-( If someone wanted to experiment with extending EZConfig to run in compile time as a quasiquoter, I'd support them. :-)

Reply:
There is a function to do sanity checks, but it has to be invoked manually in `startupHook` at present.

Reply:
> There're still people who don't use EZConfig but might still want to use modal keybindings (sooner or later), like me.

I meant to provide an interface so users can supply keybindings in EZConfig style *and* the `-> Map` style, not to replace it

Reply:
As we talked in the IRC: we'll handle that at a later point, let's merge the PR as-is for now